domain/application/state/application.go

```diff
@@ -1733,10 +1733,116 @@ INSERT INTO cloud_service (*) VALUES ($cloudService.*)
 	return nil
 }
 
+// GetApplicationStatus looks up the status of the specified application,
+// returning an error satisfying [applicationerrors.ApplicationNotFound] if the
+// application is not found.
+func (st *State) GetApplicationStatus(ctx context.Context, appID coreapplication.ID) (*application.StatusInfo[application.WorkloadStatusType], error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, jujuerrors.Trace(err)
+	}
+
+	identID := applicationID{ID: appID}
+	query, err := st.Prepare(`
+SELECT &applicationStatusInfo.*
+FROM application_status
+WHERE application_uuid = $applicationID.uuid;
+`, identID, applicationStatusInfo{})
+	if err != nil {
+		return nil, jujuerrors.Trace(err)
+	}
+	var status applicationStatusInfo
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if err := st.checkApplicationExists(ctx, tx, identID); err != nil {
+			return jujuerrors.Trace(err)
+		}
+		if err := tx.Query(ctx, query, identID).Get(&status); errors.Is(err, sqlair.ErrNoRows) {
+			// If the application status is not set, then it's up to the
+			// the caller to either return that information or use derive
+			// the status from the units.
+			return nil
+		} else if err != nil {
+			return jujuerrors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	statusType, err := decodeWorkloadStatus(status.StatusID)
+	if err != nil {
+		return nil, jujuerrors.Trace(err)
+	}
+
+	return &application.StatusInfo[application.WorkloadStatusType]{
+		Status:  statusType,
+		Message: status.Message,
+		Data:    status.Data,
+		Since:   status.UpdatedAt,
+	}, nil
+}
+
+// SetApplicationStatus saves the given application status, overwriting any
+// current status data. If returns an error satisfying
+// [applicationerrors.ApplicationNotFound] if the application doesn't exist.
+func (st *State) SetApplicationStatus(
+	ctx context.Context,
+	applicationID coreapplication.ID,
+	status *application.StatusInfo[application.WorkloadStatusType],
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return jujuerrors.Trace(err)
+	}
+
+	statusID, err := encodeWorkloadStatus(status.Status)
+	if err != nil {
+		return jujuerrors.Trace(err)
+	}
+
+	statusInfo := applicationStatusInfo{
+		ApplicationID: applicationID,
+		StatusID:      statusID,
+		Message:       status.Message,
+		Data:          status.Data,
+		UpdatedAt:     status.Since,
+	}
+	stmt, err := st.Prepare(`
+INSERT INTO application_status (*) VALUES ($applicationStatusInfo.*)
+ON CONFLICT(application_uuid) DO UPDATE SET
+    status_id = excluded.status_id,
+    message = excluded.message,
+    updated_at = excluded.updated_at,
+    data = excluded.data;
+`, statusInfo)
+	if err != nil {
+		return jujuerrors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if err := tx.Query(ctx, stmt, statusInfo).Run(); internaldatabase.IsErrConstraintForeignKey(err) {
```

Comment:
I generally prefer using a select statement first (while it is still a read transaction) to explicitly check if the required entities exists rather than checking for the error after the fact.

Reply:
I would too, but for status, I think it's fine in this one circumstance.